### PR TITLE
[MRG] Fix anonymization in EDF files to be EDF/EDFBrowser-compliant

### PR DESCRIPTION
**Pull request by adam2392**
PR Description
--------------

Closes: #667 

Working on this + unit tests.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

File: doc/whats_new.rst [hunk 1]
```diff
@@ -46,7 +46,7 @@ Requirements
 Bug fixes
 ^^^^^^^^^
 
-- xxx
+- Anonymization of EDF files in :func:`write_raw_bids` will now convert recording date to ``01-01-1985 00:00:00``, while setting the recording date in the ``scans.tsv`` file to an anonymized date < 1925, thus making the file EDF/EDFBrowser compliant, by `Adam Li`_ (:gh:`669`)
```
Comment (hoechenberger) [inline doc/whats_new.rst:49]: In fact the user has to specify the time shift in days to move the date into the correct, BIDS-compatible region (1900–1925 IIRC). So this is **not** done automatically.

File: mne_bids/read.py [hunk 2]
```diff
@@ -185,6 +186,40 @@ def _handle_participants_reading(participants_fname, raw,
...
+        show_warning = False
+
+    # acquisition time in scans.tsv and recording should match in general
+    if acq_time != raw.info['meas_date'] and show_warning:
+        warn(f'Scans.tsv acq_time {acq_time} does not match '
+             f'the recording date in the raw file '
+             f'{raw.info["meas_date"]}. '
+             f'Using the date in scans.tsv file as ground truth.')
+
+    raw.set_meas_date(acq_time)
```
Comment (hoechenberger) [inline mne_bids/read.py:218]: Really, let's drop all that special handling which is just there to ensure that metadata between raw and sidecar files matches. We'll soon push for a change / discussion in the BIDS community to drop this assumption, as we cannot properly adhere to it in some cases (like EDF anonymization) anyway.
```suggestion
    raw.set_meas_date(acq_time)
```

File: mne_bids/write.py [hunk 3]
```diff
@@ -1009,6 +1009,10 @@ def write_raw_bids(raw, bids_path, events_data=None,
             recording date will be overwritten as well. If True, keep subject
             information apart from the recording date.
 
+    convert : bool
+        Whether to convert files to RECOMMENDED BIDS format for that
+        datatype. For example, if iEEG, then the RECOMMENDED format is
+        BrainVision, so files would be converted to have extension ``.vhdr``.
```
Comment (hoechenberger) [inline mne_bids/write.py:1015]:
```suggestion
        Whether to convert files to the recommended BIDS format for that
        datatype. For example, for iEEG the recommended format is
        BrainVision, so the data would be converted to ``.vhdr`` files.
```

File: mne_bids/write.py [hunk 4]
```diff
@@ -1248,13 +1251,22 @@ def write_raw_bids(raw, bids_path, events_data=None,
...
             'Please set overwrite to True.')
 
+    # if user wants to convert data to
+    if convert:
+        if bids_path.datatype == 'meg':
+            bids_path.update(extension='.fif')
+        elif bids_path.datatype == 'eeg':
+            bids_path.update(extension='.vhdr')
+        elif bids_path.datatype == 'ieeg':
+            bids_path.update(extension='.vhdr')
```
Comment (hoechenberger) [inline mne_bids/write.py:1261]: `else` raise an exception, I suppose?

File: mne_bids/write.py [hunk 5]
```diff
@@ -1248,13 +1251,22 @@ def write_raw_bids(raw, bids_path, events_data=None,
...
     # if current format not BIDS compliant
     if not convert:
         convert = ext not in ALLOWED_DATATYPE_EXTENSIONS[bids_path.datatype]
 
-    if bids_path.datatype == 'meg' and convert and not anonymize:
-        raise ValueError(f"Got file extension {convert} for MEG data, "
+    if bids_path.datatype == 'meg' and not convert and not anonymize:
+        raise ValueError(f"Got file extension {ext} for MEG data, "
                          f"expected one of "
                          f"{ALLOWED_DATATYPE_EXTENSIONS['meg']}")
```
Comment (hoechenberger) [inline mne_bids/write.py:1271]:
```suggestion
                         f"{', '.join(sorted(ALLOWED_DATATYPE_EXTENSIONS['meg']))}")
```

File: mne_bids/write.py (hunk 3 above)
Comment (adam2392) [reply mne_bids/write.py:1015]: Sorry this is meant to address a different PR... Idk how this wound up here :/

File: mne_bids/read.py [hunk 6]
```diff
@@ -185,6 +186,28 @@ def _handle_participants_reading(participants_fname, raw,
...
+    data_fname = op.join(bids_path.datatype, bids_path.fpath.name)
+    fnames = scans_tsv['filename']
+    acq_times = scans_tsv['acq_time']
+    row_ind = fnames.index(data_fname)
+
+    # extract the acquisition time from scans file
+    acq_time = datetime.strptime(acq_times[row_ind], '%Y-%m-%dT%H:%M:%S')
+    acq_time.astimezone(tz=timezone.utc)
+
+    if verbose:
```
Comment (hoechenberger) [inline mne_bids/read.py:203]: Please use the logger instead.

File: mne_bids/read.py [hunk 7]
```diff
@@ -185,6 +186,28 @@ def _handle_participants_reading(participants_fname, raw,
     return raw
 
 
+def _handle_scans_reading(scans_fname, raw, bids_path, verbose=False):
+    """Read associated scans.tsv and set meas_date."""
+    scans_tsv = _from_tsv(scans_fname)
+
+    # get the row
+    data_fname = op.join(bids_path.datatype, bids_path.fpath.name)
```
Comment (hoechenberger) [inline mne_bids/read.py:194]: Will this work as intended on Windows?

File: mne_bids/copyfiles.py [hunk 8]
```diff
@@ -456,9 +456,11 @@ def copyfile_edf(src, dest, anonymize=None):
 
         # Try to anonymize the recording date
         daysback, keep_his = _check_anonymize(anonymize, raw, '.edf')
-        info = anonymize_info(raw.info, daysback=daysback, keep_his=keep_his)
-        startdate = datetime.strftime(info['meas_date'], '%d-%b-%Y').upper()
-        meas_date = datetime.strftime(info['meas_date'], '%d.%m.85')
+        # info = anonymize_info(raw.info, daysback=daysback, keep_his=keep_his)
+        # startdate = datetime.strftime(info['meas_date'], '%d-%b-%Y').upper()
+        # meas_date = datetime.strftime(info['meas_date'], '%d.%m.85')
+        startdate = datetime.strptime('01-01-1985', '%d-%b-%Y')
```
Comment (hoechenberger) [inline mne_bids/copyfiles.py:462]: For consistency, this should be called`start_date`

File: mne_bids/copyfiles.py [hunk 9]
```diff
@@ -456,9 +456,11 @@ def copyfile_edf(src, dest, anonymize=None):
 
         # Try to anonymize the recording date
         daysback, keep_his = _check_anonymize(anonymize, raw, '.edf')
-        info = anonymize_info(raw.info, daysback=daysback, keep_his=keep_his)
-        startdate = datetime.strftime(info['meas_date'], '%d-%b-%Y').upper()
-        meas_date = datetime.strftime(info['meas_date'], '%d.%m.85')
+        # info = anonymize_info(raw.info, daysback=daysback, keep_his=keep_his)
+        # startdate = datetime.strftime(info['meas_date'], '%d-%b-%Y').upper()
+        # meas_date = datetime.strftime(info['meas_date'], '%d.%m.85')
```
Comment (hoechenberger) [inline mne_bids/copyfiles.py:461]: Cruft?

File: mne_bids/read.py (hunk 7 above)
Comment (adam2392) [reply mne_bids/read.py:194]: Not sure... what is the suggested way to do this? for 

`eeg/sub<blah>_eeg.edf`?

File: mne_bids/tests/test_write.py [hunk 10]
```diff
@@ -2117,29 +2120,56 @@ def test_undescribed_events(_bids_validate, drop_undescribed_events):
     _bids_validate(bids_root)
 
 
-def test_anonymize_empty_room():
```
Comment (adam2392) [inline mne_bids/tests/test_write.py:2120]: turned this into a parameterized test for multiple file types

File: mne_bids/read.py (hunk 7 above)
Comment (adam2392) [reply mne_bids/read.py:194]: Noticed CI has a windows, so I'll wait to see if it get's handled.

**Comment by adam2392** [comment]
There looks to be some failing examples:

```
Unexpected failing examples:
/home/circleci/project/examples/read_bids_datasets.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/project/examples/read_bids_datasets.py", line 154, in <module>
    raw = read_raw_bids(bids_path=bids_path, verbose=False)
  File "/home/circleci/project/mne_bids/read.py", line 520, in read_raw_bids
    raw = _handle_scans_reading(scans_fname, raw, bids_path,
  File "/home/circleci/project/mne_bids/read.py", line 214, in _handle_scans_reading
    acq_time = datetime.strptime(acq_time, '%Y-%m-%dT%H:%M:%S.%fZ')
  File "/usr/local/lib/python3.8/_strptime.py", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
  File "/usr/local/lib/python3.8/_strptime.py", line 349, in _strptime
    raise ValueError("time data %r does not match format %r" %
ValueError: time data '2007-07-05T13:17:11' does not match format '%Y-%m-%dT%H:%M:%S.%fZ'


/home/circleci/project/examples/update_bids_datasets.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/project/examples/update_bids_datasets.py", line 100, in <module>
    raw = read_raw_bids(bids_path=bids_path)
  File "/home/circleci/project/mne_bids/read.py", line 520, in read_raw_bids
    raw = _handle_scans_reading(scans_fname, raw, bids_path,
  File "/home/circleci/project/mne_bids/read.py", line 214, in _handle_scans_reading
    acq_time = datetime.strptime(acq_time, '%Y-%m-%dT%H:%M:%S.%fZ')
  File "/usr/local/lib/python3.8/_strptime.py", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
  File "/usr/local/lib/python3.8/_strptime.py", line 349, in _strptime
    raise ValueError("time data %r does not match format %r" %
ValueError: time data '2007-07-05T13:17:11' does not match format '%Y-%m-%dT%H:%M:%S.%fZ'
```

This is probably due to the fact that certain datasets with written acq_time don't have the microseconds encoded. Fixing now.

**Comment by adam2392** [comment]
This is good to go from my end. The failing CI is due to pulling data from mne-testing-data.

File: mne_bids/read.py [hunk 11]
```diff
@@ -185,6 +186,45 @@ def _handle_participants_reading(participants_fname, raw,
...
+        bids_path.update(split=None, extension=extension)
+        fname = bids_path.basename
+    elif fname.endswith('.pdf'):
+        # for BTI files, the scan is an entire directory
+        fname = fname.split('.')[0]
+
+    # get the row corresponding to the file
+    # use string concatenation instead of os.path
+    # to work nicely with windows
+    data_fname = bids_path.datatype + '/' + fname
```
Comment (sappelhoff) [inline mne_bids/read.py:206]: wouldn't this work just as well?

```suggestion
    data_fname = bids_path.datatype / fname
```

on Linux, it does:
```Python
In [6]: a =pathlib.Path("hihho")

In [7]: b= "nono"

In [8]: a / b
Out[8]: PosixPath('hihho/nono')

```

**Review by sappelhoff — COMMENTED** [review]
LGTM overall. @a-hurst what do you think?

File: mne_bids/read.py (hunk 7 above)
Comment (sappelhoff) [reply mne_bids/read.py]: why wouldn't `bids_path.datatype / bids_path.fpath.name` work? I thought that's one of the main syntactic sugars of pathlib.Path?

**Review by hoechenberger — APPROVED** [review]

File: mne_bids/read.py (hunk 11 above)
Comment (hoechenberger) [reply mne_bids/read.py:206]: @sappelhoff  but `bids_path.datatype` is a string

**Comment by hoechenberger** [comment]
@adam2392 Do you think you could try to increase the coverage? Slightly?

https://codecov.io/gh/mne-tools/mne-bids/pull/669/src/mne_bids/write.py

File: mne_bids/read.py (hunk 7 above)
Comment (sappelhoff) [reply mne_bids/read.py]: Okay, I see - thanks!

**Comment by adam2392** [comment]
> @adam2392 Do you think you could try to increase the coverage? Slightly?
> 
> https://codecov.io/gh/mne-tools/mne-bids/pull/669/src/mne_bids/write.py

The link doesn't lead anywhere for me :/. 

But I'm guessing it might be the missing line in `_handle_scans_reading` regarding:

```
        # microseconds in the acquisition time is optional
        if '.' not in acq_time:
            # acquisition time ends with '.%fZ' microseconds string
            acq_time += '.0Z'
```

since this is only tested in the examples. I'll add a unit test though.

**Comment by sappelhoff** [comment]
I think also this one:

image

**Comment by adam2392** [comment]
> I think also this one:
> 
> image

Aren't all these lines covered (green color)? Sorry if not, I'm not 100% clear on codecov UI sometimes.

**Comment by sappelhoff** [comment]
The red lines are not covered

**Comment by adam2392** [comment]
> The red lines are not covered ... Oh, now I am confused. It's red and green. Idk

Just looked it up again to remind myself too :p. 

So I believe the red lines w/ the ``-`` sign indicate the lines were removed from VC (I removed those). The green lines indicate covered lines. Unless... that ``ValueError`` is not covered, which would be weird since I didn't change any tests there?

**Comment by sappelhoff** [comment]
This is what the report says about your changes:

image


This is the "legend" of what codecov says the output means:

image


So your commit diff for write.py is 0% covered, apparently.

This is the part that you changed:

image

The red block indicates that the valueerror raising is NOT covered.

I think the issue is that that part was also not covered before ... however since you adjusted the lines, codecov now expects you to cover the diff ... i.e., codecov doesn't care that the line was also uncovered before.

--> at least it hink that's what's going on.

**Comment by adam2392** [comment]
> This is what the report says about your changes:
> 
> image
> 
> This is the "legend" of what codecov says the output means:
> 
> image
> 
> So your commit diff for write.py is 0% covered, apparently.
> 
> This is the part that you changed:
> 
> image
> 
> The red block indicates that the valueerror raising is NOT covered.
> 
> I think the issue is that that part was also not covered before ... however since you adjusted the lines, codecov now expects you to cover the diff ... i.e., codecov doesn't care that the line was also uncovered before.
> 
> --> at least it hink that's what's going on.

Okay, I think the issue here is that there is no "unsupported" MEG mne-reader in BIDS that I know of? All the ctf, kit, bti and fif are all supported.  I think this error was put in as a catch-all in case of other MEG manufacturers.

I added in an in-line comment for now.

**Comment by adam2392** [comment]
These lines:

```
# Anonymize
    convert = False
    if anonymize is not None:
        daysback, keep_his = _check_anonymize(anonymize, raw, ext)
        raw.anonymize(daysback=daysback, keep_his=keep_his, verbose=verbose)

        if bids_path.datatype == 'meg' and ext != '.fif':
            if verbose:
                warn('Converting to FIF for anonymization')
            convert = True
            bids_path.update(extension='.fif')
        elif bids_path.datatype in ['eeg', 'ieeg']:
            if ext not in ['.vhdr', '.edf', '.bdf']:
                if verbose:
                    warn('Converting data files to BrainVision format '
                         'for anonymization')
                convert = True
                bids_path.update(extension='.vhdr')
```

followed by:

```
# If not already converting for anonymization, we may still need to do it
    # if current format not BIDS compliant
    if not convert:
        convert = ext not in ALLOWED_DATATYPE_EXTENSIONS[bids_path.datatype]

    # check if there
    if bids_path.datatype == 'meg' and convert and not anonymize:
        raise ValueError(
            f"Got file extension {convert} for MEG data, "
            f"expected one of "
            f"{', '.join(sorted(ALLOWED_DATATYPE_EXTENSIONS['meg']))}")
```

makes that this ValueError isn't reached according to my walk-through.

**Comment by adam2392** [comment]
Not sure why tests are failing on CI, since they work locally for me.

```
>>> sys_info()
Platform:      macOS-10.16-x86_64-i386-64bit
Python:        3.8.5 | packaged by conda-forge | (default, Jul 22 2020, 17:24:51)  [Clang 10.0.0 ]
Executable:    /Users/adam2392/opt/miniconda3/envs/mne/bin/python
CPU:           i386: 8 cores
Memory:        Unavailable (requires "psutil" package)
mne:           0.23.dev0
numpy:         1.19.1 {blas=NO_ATLAS_INFO, lapack=lapack}
scipy:         1.5.2
matplotlib:    3.3.0 {backend=TkAgg}

sklearn:       0.24.0
numba:         Not found
nibabel:       3.2.1
nilearn:       0.7.0
dipy:          Not found
cupy:          Not found
pandas:        1.1.3
mayavi:        Not found
pyvista:       0.26.1 {pyvistaqt=0.2.0, OpenGL 4.1 ATI-4.2.13 via AMD Radeon Pro 560 OpenGL Engine}
vtk:           9.0.1
PyQt5:         5.15.1
```

**Comment by agramfort** [comment]
I can replicate the pb with mne 0.21.2 on my mac

**Comment by adam2392** [comment]
> I can replicate the pb with mne 0.21.2 on my mac

Okay it turns out that somewhere between mne0.21.2 and mne0.22, this was fixed, so that one could set a channel as `mag`. I refactored the test, and now it should work.

**Comment by agramfort** [comment]
@adam2392 you still have 2 failures

please tell me when it's green

**Comment by adam2392** [comment]
> @adam2392 you still have 2 failures
> 
> please tell me when it's green

It's good now.

The CIs keep failing because they try to pull the mne-testing-data (i.e. flaky server connections).

**Review by agramfort — APPROVED** [review]
@sappelhoff merge if happy

**Review by sappelhoff — APPROVED** [review]
thanks adam, I think this is the best we can do given the BIDS and EDF constraints.

let's merge and hope that we'll loosen up the BIDS constraints soon.

**Pull request merged**